### PR TITLE
test: SLT test with queries were all parts of the query are essential

### PR DESCRIPTION
**Pull request by philip-stoev**
Out of a randomized TPC- like workload, we have extracted as an SLT
test queries where each part of the query is essential to producing
the correct result.

If any part or predicate of such a query is mis-optimized, mis-transformed
or mis-evaluated, the test will fail.

### Motivation

The RQG, the fuzz testing framework that I am using, remains somewhat problematic to run in an automated context, as it will report various errors as it runs and those errors need to be seen by a human to decide if they constitute a regression or not.

While I will be working to containerize it and put the RQG tests directly in the CI, the immediate need remains to add more tests against the Optimizer/Dataflow that are runnable without invoking the RQG.

One solution to this would be to extract interesting queries out of all those generated by the RQG and put them in SLT and .td so that they run like any other test.

One way to extract "interesting" queries is to fish for queries where every part of the query is essential, that is, pick queries where accidentally dropping, mis-transforming or mis-evaluating any particular SQL keyword, predicate or part of a predicate would cause in the result of the final query to become different, thus exposing the bug.

This particular approach is quite simplistic and only as good as the input of generated queries -- in this case, only one query with an outer join was emitted into the SLT test. Some other SQL constructs that were present in the initial SQL grammar that the generation started from did not make it at all in the final list.

The purpose of this PR is essentially to allow the team to discuss if we want such autogenerated tests in the codebase, where should they sit, etc. before I continue to add more, generated using various grammars and picked using various strategies.

For example, should this be agreed to be a valid testing approach, a more complicated picking algorithm would be used whereby all input queries are ranked according to various fitness measures and the best representatives are chosen for inclusion in an SLT test.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/9747)

### CODE REVIEW

**Comment by philip-stoev** [comment]
@benesch how can I prevent doctest from running the scripts being linted? 

```
$ bin/pyactivate --dev -m pytest -qq --doctest-modules misc/python
--
  |  
  | ==================================== ERRORS ====================================
  | ______ ERROR collecting materialize/query_pickers/all_parts_essential.py _______
  | misc/python/materialize/query_pickers/all_parts_essential.py:29: in <module>
  | conn = psycopg2.connect("dbname=pstoev user=pstoev")
  | /usr/local/lib/python3.8/dist-packages/psycopg2/__init__.py:122: in connect
  | conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
  | E   psycopg2.OperationalError: could not connect to server: No such file or directory
  | E   	Is the server running locally and accepting
  | E   	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?
```

**Review by benesch — COMMENTED** [review]
In Python imports are not meant to have side-effects (see e.g. https://news.ycombinator.com/item?id=7536246), so module-level variables that e.g. open a database connection are a no-no. In this case I'd recommend creating the `conn` in the `run` method and then passing the `conn` or `cursor` object on the fly to each method that needs it.

File: test/sqllogictest/autogenerated_all_parts_essential.slt [hunk 1]
```diff
@@ -0,0 +1,994 @@
...
+
+%10 =
+| Union %8 %9
+
+EOF
+
+query II
+SELECT c_custkey , o_custkey FROM lineitem JOIN orders ON ( l_orderkey = o_orderkey ) JOIN customer ON ( c_acctbal <= o_totalprice ) WHERE l_quantity = 17 AND c_comment BETWEEN 's' AND 'x' OR l_quantity  IN ( 7 , 33 )
+----
+1  1
```
Comment (wangandi) [inline test/sqllogictest/autogenerated_all_parts_essential.slt]: Expected results should be in "values hashing to" form. See https://github.com/MaterializeInc/materialize/pull/8392#issuecomment-925299207

**Review by wangandi — COMMENTED** [review]
We already have full SLT for detecting wrong results. Do we have a sense of whether it would be possible to try and autogenerate queries that cover spaces that full SLT does not?

**Comment by wangandi** [comment]
Can the autogenerated files go in a new folder `test/sqllogictest/autogenerate` just to make it easier to browse through the test/sqllogictest folder?

**Comment by benesch** [comment]
If the autogenerated files are gonna get pretty large, we probably should split them out into a submodule like we do for the sqlite sqllogictests. That makes updating/editing them painful, but IME it's a cost worth paying. Once you've committed a large file to your repository, there's basically no undoing that ever, and that's a cost that everyone who checks out our source has to pay forever.

File: test/sqllogictest/autogenerated_all_parts_essential.slt [hunk 2]
```diff
@@ -0,0 +1,994 @@
...
+statement ok
+CREATE INDEX pk_customer_custkey ON customer (c_custkey);
+
+statement ok
+CREATE INDEX fk_customer_nationkey ON customer (c_nationkey ASC);
+
+statement ok
+CREATE TABLE orders (
+    o_orderkey       integer,
+    o_custkey        integer NOT NULL,
```
[the remark this answers is not in the document]
Comment (philip-stoev) [reply test/sqllogictest/autogenerated_all_parts_essential.slt]: This schema and the generator that comes with it does not support NULL-able and NULL values. However, I have another schema and data that has those and I will be generating queries against that schema+data separately.

File: test/sqllogictest/autogenerated_all_parts_essential.slt (hunk 1 above)
Comment (philip-stoev) [reply test/sqllogictest/autogenerated_all_parts_essential.slt]: Yes, thanks for reminding me that this feature exists -- I was struggling with some queries that returned large result sets.

**Comment by philip-stoev** [comment]
> Can the autogenerated files go in a new folder `test/sqllogictest/autogenerate` just to make it easier to browse through the test/sqllogictest folder?

Yes, most definitely, I was just about to do that.

**Comment by philip-stoev** [comment]
> If the autogenerated files are gonna get pretty large, we probably should split them out into a submodule like we do for the sqlite sqllogictests. That makes updating/editing them painful, but IME it's a cost worth paying. Once you've committed a large file to your repository, there's basically no undoing that ever, and that's a cost that everyone who checks out our source has to pay forever.

The idea is for them to not get large, especially when the hash of the results is recorded and not the actual results.

Instead of committing 1M queries to the repository, as some third-party SLT tests have done, commit only 10 or so queries per test that are by some metrics the best queries out of the larger set of autogenerated queries you are starting with.

So we are hopefully talking about:
- dozens of test files
- at most 100K bytes per file

Let me try the submodule idea and see how much developer overhead there is when updating the tests.

**Comment by philip-stoev** [comment]
> We already have full SLT for detecting wrong results. Do we have a sense of whether it would be possible to try and autogenerate queries that cover spaces that full SLT does not?

Yes, theoretically it would be possible to rank queries according to an abstract "covers stuff that SLT does not" fitness metric that could be comprised of the following:
- code coverage touched by the autogenerated query that is not touched by the SLT queries
- AST constructs present in the autogenerated query that are not present in the SLT queries
- query plan constructs present in the autogenerated query that are not present in the SLT

The code coverage part is a bit problematic in Rust, but in general it is known that it is possible to code a tool that for a given query will give you the code coverage report for this particular query. This code coverage report can then be diced and sliced to decide how useful the query is, e.g. it can be intersected with the combined coverage report from all SLT queries to determine if the autogenerated query provides any unique code coverage contribution.

The AST / query plan fitness would be I guess some sort of an NLP problem -- if we consider the SQL or the EXPLAIN to be a "sentence", then all the SLT queries will form a "corpus" and then what we will be looking for is autogenerated queries that score high on some dissimularity metric against that corpus. Something like Amazon's "Statistically improbable phrase" I guess...

**Review by frankmcsherry — APPROVED** [review]
Seems good to me, but probably many opinions!

File: test/sqllogictest/autogenerated_all_parts_essential.slt (hunk 1 above)
Comment (philip-stoev) [reply test/sqllogictest/autogenerated_all_parts_essential.slt]: Done, plus a fix to support hashed results with --rewrite-results in https://github.com/MaterializeInc/materialize/pull/9998

**Comment by philip-stoev** [comment]
The sqlite autogenerated tests are 1Mb+ each, and hundreds of them. They also do not contain any EXPLAIN plans, so their contents are decoupled from the particular revision of the Mz optimizer.

By contrast, the autogenerated tests envisioned are 100K bytes in size, which is comparable in magnitude to other SLT tests we already have in the repository, and contain EXPLAIN plans, which may need to change frequently.

So I will revisit the submodule idea after we have accumulated a few of those files and we have been able to establish whether they are generally useful and how much pain it is to maintain them.

**Pull request merged**